pkg/ipcache/metadata.go

```diff
@@ -104,114 +151,178 @@ func (m *metadata) get(prefix string) labels.Labels {
-		// cluster), or CIDR IDs for kube-apiservers deployed outside of the
-		// cluster.
-		// Also, any new identity should be upserted, regardless.
-		if hasKubeAPIServerLabel || isNew {
-			tmpSrc := src
-			if hasKubeAPIServerLabel {
-				// Overwrite the source because any IP associated with the
-				// kube-apiserver takes the strongest precedence. This is
-				// because we need to overwrite Local if only the local node IP
-				// has been upserted into the ipcache first.
+			// Insert to propagate the updated set of labels after removal.
+			newID, _, err = ipc.injectLabels(ctx, prefix, lbls)
+			if err != nil {
+				// NOTE: This may fail during a 2nd or later
+				// iteration of the loop. To handle this, break
+				// the loop here and continue executing the set
+				// of changes for the prefixes that were
+				// already processed.
 				//
-				// Also, trigger policy recalculations to update kube-apiserver
-				// identity.
-				newLbls := id.Labels
-				tmpSrc = source.KubeAPIServer
-				trigger = true
-				// If host identity has changed, update its labels.
-				if id.ID == identity.ReservedIdentityHost {
-					identity.AddReservedIdentityWithLabels(id.ID, newLbls)
-				}
-				idsToPropagate[id.ID] = newLbls.LabelArray()
-			}
-
-			toUpsert[prefix] = Identity{
-				ID:     id.ID,
-				Source: tmpSrc,
-			}
-		} else {
-			// Unlikely, but to balance the allocation / release
-			// we must either add the identity to `toUpsert`, or
-			// immediately release it again. Otherwise it will leak.
-			if _, err := ipc.IdentityAllocator.Release(context.TODO(), id, false); err != nil {
+				// Old identities corresponding to earlier
+				// prefixes may be released as part of this,
+				// so hopefully this forward progress will
+				// unblock subsequent calls into this function.
 				log.WithError(err).WithFields(logrus.Fields{
-					logfields.IPAddr: prefix,
-					logfields.Labels: lbls,
-				}).Error(
-					"Failed to release assigned identity during label injection, this might be a leak.",
-				)
-			} else {
-				log.WithFields(logrus.Fields{
 					logfields.IPAddr:   prefix,
 					logfields.Identity: id,
-				}).Debug(
-					"Released identity to balance reference counts",
+					logfields.Labels:   lbls, // new labels
+				}).Warning(
+					"Failed to allocate new identity while handling change in labels associated with a prefix.",
 				)
+				remainingPrefixes = modifiedPrefixes[i:]
+				err = fmt.Errorf("failed to allocate new identity during label injection: %w", err)
+				break
+			}
+
+			// It's plausible to pull the same information twice
+			// from different sources, for instance in etcd mode
+			// where node information is propagated both via the
+			// kvstore and via the k8s control plane. If the new
+			// security identity is the same as the noe currently
+			// being used, then no need to update it.
+			if id.ID == newID.ID {
+				goto releaseIdentity
+			}
```

Comment:
Isn't that what `isNew` can tell us? `isNew` comes from the second return from `injectLabels` which we are ignoring currently.

Also, is it necessary to have this logic inside the IDMD? I'm thinking why can't we just `UpsertLabels` from the kvstore side and have the IDMD calculate that it doesn't need to do anything because we have the same information and from a lower prio source?

Reply:
I don't think `isNew` is quite the same. Let's take the case where we have an IP -> `remote-node` mapping. Then kube-apiserver becomes scheduled on that node. We will insert this information to the metadata map and trigger this function. When attempting to determine the new identity, it should be updated to the reserved kube-apiserver identity (7). In this case, the identity is a reserved identity so `isNew` is false. However, we do expect that the identity is different from before and the ipcache datapath should be updated. Checking `isNew` here would prevent this from updating correctly.

We'll need to be a bit careful when we extend this code to handle encryption keys better though, since we could end up having the same labels but need to update the encryption key. Maybe the generalized version of this check is "check the current BPF ipcache value format and compare to the new BPF ipcache value format, then only push it if it changed".

I'll have to think a bit more about the latter question. One concern is that if we have multiple sources calling `UpsertLabels()`, source should *not* be considered for determining the value of the entries any more. Part of the point of the PR is to allow multiple sources of labels to be combined. There's also the question of where we write the logic to determine exactly how these updates are pushed to the ipcache. Currently we emit events for all changes into the metadata map, and all logic for processing those changes is in this function. It may make sense to push an optimization pass to avoid calls into this function via such checks at some point. Right now I'm not too sure it will make a big difference, especially given that these identities are allocated locally anyway.

Reply:
> In this case, the identity is a reserved identity so isNew is false. However, we do expect that the identity is different from before and the ipcache datapath should be updated. Checking isNew here would prevent this from updating correctly.

Ah yes, I forgot about this :+1: 

> generalized version of this check is "check the current BPF ipcache value and compare to the new BPF ipcache value after this change is processed, then only push the ipcache entry into the map it if it changed"

Agree, this is likely what we need to do to associate aux information with IPs. 

---

For the latter, I was thinking a diff like below.

```diff
diff --git a/pkg/ipcache/kvstore.go b/pkg/ipcache/kvstore.go
index e327b06d19..b3564e518a 100644
--- a/pkg/ipcache/kvstore.go
+++ b/pkg/ipcache/kvstore.go
@@ -17,7 +17,9 @@ import (
 	"github.com/sirupsen/logrus"
 
 	"github.com/cilium/cilium/pkg/identity"
+	"github.com/cilium/cilium/pkg/ipcache/types"
 	"github.com/cilium/cilium/pkg/kvstore"
+	"github.com/cilium/cilium/pkg/labels"
 	"github.com/cilium/cilium/pkg/lock"
 	"github.com/cilium/cilium/pkg/logging/logfields"
 	"github.com/cilium/cilium/pkg/option"
@@ -221,6 +223,7 @@ func NewIPIdentityWatcher(ipc *IPCache, backend kvstore.BackendOperations) *IPId
 func (iw *IPIdentityWatcher) Watch(ctx context.Context) {
 
 	var scopedLog *logrus.Entry
+	rid := types.NewResourceID(types.ResourceKindKVStore, "", "")
 restart:
 	watcher := iw.backend.ListAndWatch(ctx, "endpointIPWatcher", IPIdentitiesPath, 512)
 
@@ -295,15 +298,15 @@ restart:
 					}
 				}
 
-				peerIdentity := ipIDPair.ID
-				if option.Config.EnableRemoteNodeIdentity && peerIdentity == identity.ReservedIdentityHost {
+				peerLabel := labels.LabelHost
+				if option.Config.EnableRemoteNodeIdentity && ipIDPair.ID == identity.ReservedIdentityHost {
 					// The only way we can discover IPs associated with the local host
 					// is directly via the NodeDiscovery package. If someone is informing
 					// this agent about IPs corresponding to the "host" via the kvstore,
 					// then they're sharing their own perspective on their own node IPs'
 					// identity. However, this node has remote-node enabled, so we should
 					// treat the peer as a "remote-node", not a "host".
-					peerIdentity = identity.ReservedIdentityRemoteNode
+					peerLabel = labels.LabelRemoteNode
 				}
 
 				// There is no need to delete the "old" IP addresses from this
@@ -312,10 +315,7 @@ restart:
 				// and the endpoint-runIPIdentitySync where it bounded to a
 				// lease and a controller which is stopped/removed when the
 				// endpoint is gone.
-				iw.ipcache.Upsert(ip, ipIDPair.HostIP, ipIDPair.Key, k8sMeta, Identity{
-					ID:     peerIdentity,
-					Source: source.KVStore,
-				})
+				iw.ipcache.UpsertLabels(ip, peerLabel, source.KVStore, rid)
 
 			case kvstore.EventTypeDelete:
 				// Value is not present in deletion event;
diff --git a/pkg/ipcache/types/types.go b/pkg/ipcache/types/types.go
index b3aeb106e9..701f0f1a95 100644
--- a/pkg/ipcache/types/types.go
+++ b/pkg/ipcache/types/types.go
@@ -41,6 +41,7 @@ var (
 	ResourceKindEndpointSlice        = ResourceKind("endpointslices")
 	ResourceKindEndpointSlicev1beta1 = ResourceKind("endpointslices_v1beta1")
 	ResourceKindNode                 = ResourceKind("node")
+	ResourceKindKVStore              = ResourceKind("kvstore")
 )
 
 // NewResourceID returns a ResourceID populated with the standard fields for
```

Reply:
AFAIK ipcache/kvstore.go will receive both node and pod addresses, so we can't just assume that the identity is for a node. Specifically this diff will ignore any other identity incorrectly:

```
-				peerIdentity := ipIDPair.ID
...
+				peerLabel := labels.LabelHost
```

Reply:
I think you're hoping to turn the kvstore notifications for other nodes into label metadata in the metadata cache so that it properly merges with kube-apiserver label? Maybe then we can do something special like this specifically for the host/remote-node identities.

Reply:
I have also been thinking about the pod identities and the way they should be inserted into the ipcache package overall. Background assumptions are: policy either selects endpoints by label (ie select the global identity corresponding to that endpoint), or selects peers outside the cluster (typically CIDR or FQDN policy statements, where each node comes up with its own local identity for each peer/prefix).

Given these assumptions, it is possible for a conflict between the identity associated with the IP itself vs. a local identity allocated for the same IP via CIDR/FQDN policy. Given that the former policy only applies to pods and the latter is only intended to apply to connections to peers outside the cluster, we treat the presence of a global identity for the IP as a strong signal that the peer actually lives inside the cluster. If it's inside the cluster, then the CIDR policy doesn't apply. Therefore we can ignore the local identity when there is a conflict.

Today we implement this through this weird prioritization where we allow users of IPCache to insert entries into the ipcache either using an IP directly (implying it's actually a pod) or using a prefix (which could be a /32, hence directly conflicting). Then implicitly if there is a conflict like this, we choose the identity associated with the "w.x.y.z" entry rather than the "w.x.y.z/N" key in the ipcache userspace map.

We could do a similar thing instead by inserting the prefix+identity info into the metadata map directly. When we attempt to resolve the identity for that IP as part of the regular inner loop, first check whether any of the resources provide an identity directly, and then if so, pick that identity (prioritizing the highest source first). This could be done explicitly to make the resolution rules clear. Then only if there are no resources directly providing the identity, we fall back to the current code to merge labels from all resources and create a local identity.

I don't think we have to tackle this problem in the current PR but it should be a nice incremental step on top which could lead to a point where all ipcache interactions come in using the new interface, resolved consistently using the one clear definition on how to resolve the "correct identity" for each prefix.

Reply:
Ah good call on the incorrect kvstore patch for pod identities. 

> I think you're hoping to turn the kvstore notifications for other nodes into label metadata in the metadata cache so that it properly merges with kube-apiserver label? Maybe then we can do something special like this specifically for the host/remote-node identities.

Yes this is what I was thinking.

I've updated the diff below. It still feels a bit clunky but I'm not sure there's a way around that. 

<details> 
  <summary>Diff</summary>

```diff
diff --git a/pkg/ipcache/kvstore.go b/pkg/ipcache/kvstore.go
index e327b06d19..2ac2e77b7e 100644
--- a/pkg/ipcache/kvstore.go
+++ b/pkg/ipcache/kvstore.go
@@ -17,7 +17,9 @@ import (
 	"github.com/sirupsen/logrus"
 
 	"github.com/cilium/cilium/pkg/identity"
+	"github.com/cilium/cilium/pkg/ipcache/types"
 	"github.com/cilium/cilium/pkg/kvstore"
+	"github.com/cilium/cilium/pkg/labels"
 	"github.com/cilium/cilium/pkg/lock"
 	"github.com/cilium/cilium/pkg/logging/logfields"
 	"github.com/cilium/cilium/pkg/option"
@@ -221,6 +223,7 @@ func NewIPIdentityWatcher(ipc *IPCache, backend kvstore.BackendOperations) *IPId
 func (iw *IPIdentityWatcher) Watch(ctx context.Context) {
 
 	var scopedLog *logrus.Entry
+	rid := types.NewResourceID(types.ResourceKindKVStore, "", "")
 restart:
 	watcher := iw.backend.ListAndWatch(ctx, "endpointIPWatcher", IPIdentitiesPath, 512)
 
@@ -295,28 +298,35 @@ restart:
 					}
 				}
 
-				peerIdentity := ipIDPair.ID
-				if option.Config.EnableRemoteNodeIdentity && peerIdentity == identity.ReservedIdentityHost {
-					// The only way we can discover IPs associated with the local host
-					// is directly via the NodeDiscovery package. If someone is informing
-					// this agent about IPs corresponding to the "host" via the kvstore,
-					// then they're sharing their own perspective on their own node IPs'
-					// identity. However, this node has remote-node enabled, so we should
-					// treat the peer as a "remote-node", not a "host".
-					peerIdentity = identity.ReservedIdentityRemoteNode
+				var nodePeerLabel labels.Labels
+				if ipIDPair.ID == identity.ReservedIdentityHost || ipIDPair.ID == identity.ReservedIdentityRemoteNode {
+					nodePeerLabel = labels.LabelHost
+					if option.Config.EnableRemoteNodeIdentity && ipIDPair.ID == identity.ReservedIdentityHost {
+						// The only way we can discover IPs associated with the
+						// local host is directly via the NodeDiscovery
+						// package. If someone is informing this agent about
+						// IPs corresponding to the "host" via the kvstore,
+						// then they're sharing their own perspective on their
+						// own node IPs' identity. However, this node has
+						// remote-node enabled, so we should treat the peer as
+						// a "remote-node", not a "host".
+						nodePeerLabel = labels.LabelRemoteNode
+					}
+					// There is no need to delete the "old" IP addresses from
+					// this ip ID pair. The only places where the ip ID pair
+					// are created is the clustermesh, where it sends a delete
+					// to the KVStore, and the endpoint-runIPIdentitySync where
+					// it bounded to a lease and a controller which is
+					// stopped/removed when the endpoint is gone.
+					iw.ipcache.UpsertLabels(ip, peerLabel, source.KVStore, rid)
+				} else {
+					// Identity might be a pod identity
+					iw.ipcache.Upsert(ip, ipIDPair.HostIP, ipIDPair.Key, k8sMeta, Identity{
+						ID:     ipIDPair.ID,
+						Source: source.KVStore,
+					})
 				}
 
-				// There is no need to delete the "old" IP addresses from this
-				// ip ID pair. The only places where the ip ID pair are created
-				// is the clustermesh, where it sends a delete to the KVStore,
-				// and the endpoint-runIPIdentitySync where it bounded to a
-				// lease and a controller which is stopped/removed when the
-				// endpoint is gone.
-				iw.ipcache.Upsert(ip, ipIDPair.HostIP, ipIDPair.Key, k8sMeta, Identity{
-					ID:     peerIdentity,
-					Source: source.KVStore,
-				})
-
 			case kvstore.EventTypeDelete:
 				// Value is not present in deletion event;
 				// need to convert kvstore key to IP.
diff --git a/pkg/ipcache/types/types.go b/pkg/ipcache/types/types.go
index b3aeb106e9..701f0f1a95 100644
--- a/pkg/ipcache/types/types.go
+++ b/pkg/ipcache/types/types.go
@@ -41,6 +41,7 @@ var (
 	ResourceKindEndpointSlice        = ResourceKind("endpointslices")
 	ResourceKindEndpointSlicev1beta1 = ResourceKind("endpointslices_v1beta1")
 	ResourceKindNode                 = ResourceKind("node")
+	ResourceKindKVStore              = ResourceKind("kvstore")
 )
 
 // NewResourceID returns a ResourceID populated with the standard fields for

``` 
</details>

Reply:
> We could do a similar thing instead by inserting the prefix+identity info into the metadata map directly for both cases, since the new metadata format allows us to store both types of info concurrently in the same structure. When we attempt to resolve the identity for that IP as part of the regular inner loop, first check whether any of the resources provide an identity directly, and then if so, pick that identity (prioritizing the highest source first). This could be done explicitly to make the resolution rules clear. Then only if there are no resources directly providing the identity, we fall back to the current code to merge labels from all resources and create a local identity.
>
> I don't think we have to tackle this problem in the current PR but it should be a nice incremental step on top which could lead to a point where all ipcache interactions come in using the new interface, resolved consistently using the one clear definition on how to resolve the "correct identity" for each prefix.

That sounds good to me :+1: